apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts

```diff
@@ -0,0 +1,217 @@
+    const { dappName, repo } = uploadDappInChainTGInput;
+    const gearApi = gearService.getApi();
+    const [AL] = await getAccount();
+    const sourceId = decodeAddress(AL.address);
+
+    try {
+      const response = await getLatestReleaseByRepo(
+        repo,
+        this.configService.get<string>("github.GITHUB_OWNER_REPO"),
+      );
+
+      const dapp = await this.dappDataRepository.getByNameAndRepo(
+        dappName,
+        repo,
+      );
+
+      const [optUrl, metaUrl] = this.getDownloadUrlOptAndMetaWasmByNameDapp(
+        dappName,
+        response.data.assets,
+      );
+
+      const [optWasmData, metaWasmData] = await Promise.all([
+        fetch(optUrl),
+        fetch(metaUrl),
+      ]);
+
+      const optWasmBuff = await optWasmData.buffer();
+      const metaWasmBuff = await metaWasmData.buffer();
+
+      if (dapp) {
+        return this.uploadDappInChainAndUpdate({
+          gearApi,
+          dappData: dapp,
+          optWasmBuff,
+          metaWasmBuff,
+          sourceId,
+          release: response.data.tag_name,
+        });
+      }
+      return this.uploadNewDappInChain({
+        gearApi,
+        sourceId,
+        metaWasmBuff,
+        optWasmBuff,
+        dappName,
+        release: response.data.tag_name,
+        repo,
+      });
+    } catch (error) {
+      this.logger.error(error);
+    }
+  }
+
+  public async getLatestReleasesDapps() {
+    try {
+      const res = await getAllLatestReleaseDapps();
+      console.log(res);
+    } catch (error) {
+      this.logger.error(error);
+    }
+  }
+
+  private async uploadDappInChainAndUpdate(uploadDappInChainInput: UploadDappInChainInput): Promise<DappData> {
+    const { dappData, optWasmBuff, metaWasmBuff, sourceId, gearApi, release } = uploadDappInChainInput;
+
+    const meta = await getWasmMetadata(metaWasmBuff);
+
+    const gas = await gearApi.program.calculateGas.initUpload(
+      sourceId,
+      optWasmBuff,
+      meta.types,
+      0,
+      true,
+    );
+    const program = {
+      code: optWasmBuff,
+      gasLimit: gas.min_limit,
+      value: 0,
+    };
+    const { programId } = gearApi.program.upload(program, meta);
+
+    dappData.id = programId;
+    dappData.release = release;
+    dappData.metaWasmBase64 = metaWasmBuff.toString("base64");
+
+    const updateDapp = await this.dappDataRepository.save(dappData);
+
+    return { ...updateDapp, metaWasmBase64: "long" };
+  }
+
+  private async uploadNewDappInChain(uploadNewDappInChainInput: UploadNewDappInChainInput): Promise<DappData> {
+    const { sourceId, metaWasmBuff, optWasmBuff, gearApi, dappName, release, repo } = uploadNewDappInChainInput;
+    const paylaod = this.getPayloadByDappName(dappName);
+
+    if (!paylaod) {
+      throw new BadRequestExc("Invalid dapp name");
+    }
+
+    const meta = await getWasmMetadata(metaWasmBuff);
+
+    const gas = await gearApi.program.calculateGas.initUpload(
+      sourceId,
+      optWasmBuff,
+      meta.types,
```

Comment:
Why did you add `meta.types` here? The payload should be here.

Reply:
Fixed